Changes in module 'nf-core/stitch'
--- modules/nf-core/stitch/main.nf
+++ modules/nf-core/stitch/main.nf
@@ -8,8 +8,8 @@
         'biocontainers/r-stitch:1.6.10--r43h06b5641_0' }"
 
     input:
-    tuple val(meta) , path(posfile), path(input, stageAs: "input"), path(rdata, stageAs: "RData_in"), val(chromosome_name), val(K), val(nGen)
-    tuple val(meta2), path(collected_crams), path(collected_crais), path(cramlist)
+    tuple val(meta2), path(posfile), path(input, stageAs: "input"), path(rdata, stageAs: "RData_in"), val(chromosome_name), val(K), val(nGen)
+    tuple val(meta), path(collected_crams), path(collected_crais), path(cramlist)
     tuple val(meta3), path(fasta), path(fasta_fai)
     val seed
 

************************************************************
