Changes in module 'nf-core/bcftools/view'
--- modules/nf-core/bcftools/view/main.nf
+++ modules/nf-core/bcftools/view/main.nf
@@ -30,7 +30,7 @@
                     args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
                     args.contains("--output-type z") || args.contains("-Oz") ? "vcf.gz" :
                     args.contains("--output-type v") || args.contains("-Ov") ? "vcf" :
-                    "vcf"
+                    "vcf.gz"
     """
     bcftools view \\
         --output ${prefix}.${extension} \\

************************************************************
