Changes in module 'nf-core/samtools/view'
--- modules/nf-core/samtools/view/main.nf
+++ modules/nf-core/samtools/view/main.nf
@@ -44,9 +44,11 @@
         ${reference} \\
         ${readnames} \\
         $args \\
+        ${subsample_cmd} \\
         -o ${prefix}.${file_type} \\
         $input \\
-        $args2
+        $args2 \\
+        ${region_cmd}
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
