Changes in module 'nf-core/samtools/view'
--- modules/nf-core/samtools/view/environment.yml
+++ modules/nf-core/samtools/view/environment.yml
@@ -1,8 +1,10 @@
 name: samtools_view
+
 channels:
   - conda-forge
   - bioconda
   - defaults
+
 dependencies:
+  - bioconda::htslib=1.19.1
   - bioconda::samtools=1.19.2
-  - bioconda::htslib=1.19.1

--- modules/nf-core/samtools/view/main.nf
+++ modules/nf-core/samtools/view/main.nf
@@ -8,7 +8,7 @@
         'biocontainers/samtools:1.19.2--h50ea8bc_0' }"
 
     input:
-    tuple val(meta), path(input), path(index)
+    tuple val(meta), path(input), path(index), val(region), val(subsample)
     tuple val(meta2), path(fasta)
     path qname
 
@@ -28,8 +28,10 @@
     def args = task.ext.args ?: ''
     def args2 = task.ext.args2 ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def reference = fasta ? "--reference ${fasta}" : ""
-    def readnames = qname ? "--qname-file ${qname}": ""
+    def reference      = fasta      ? "--reference ${fasta}"     : ""
+    def readnames      = qname      ? "--qname-file ${qname}"    : ""
+    def region_cmd     = region     ? "${region}"                : ""
+    def subsample_cmd  = subsample  ? "--subsample ${subsample}" : ""
     def file_type = args.contains("--output-fmt sam") ? "sam" :
                     args.contains("--output-fmt bam") ? "bam" :
                     args.contains("--output-fmt cram") ? "cram" :
@@ -42,9 +44,12 @@
         ${reference} \\
         ${readnames} \\
         $args \\
+        ${subsample_cmd} \\
         -o ${prefix}.${file_type} \\
         $input \\
-        $args2
+        $args2 \\
+        ${region_cmd}
+
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
