Changes in module 'nf-core/bcftools/pluginsplit'
--- modules/nf-core/bcftools/pluginsplit/main.nf
+++ modules/nf-core/bcftools/pluginsplit/main.nf
@@ -24,14 +24,21 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def suffix = task.ext.suffix ?: ""
 
     def samples_arg = samples ? "--samples-file ${samples}" : ""
     def groups_arg  = groups  ? "--groups-file ${groups}"   : ""
     def regions_arg = regions ? "--regions-file ${regions}" : ""
     def targets_arg = targets ? "--targets-file ${targets}" : ""
+    def extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :
+            args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
+            args.contains("--output-type z") || args.contains("-Oz") ? "vcf.gz" :
+            args.contains("--output-type v") || args.contains("-Ov") ? "vcf" :
+            "vcf"
 
     """
     bcftools plugin split \\
+        ${args} \\
         ${vcf} \\
         ${samples_arg} \\
         ${groups_arg} \\
@@ -39,7 +46,7 @@
         ${targets_arg} \\
         --output ${prefix}
 
-    mv ${prefix}/* .
+    for i in ${prefix}/*; do cp "\$i" "./\$(basename "\$i" .${extension})${suffix}.${extension}"; done
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
