Changes in module 'nf-core/glimpse2/phase'
--- modules/nf-core/glimpse2/phase/main.nf
+++ modules/nf-core/glimpse2/phase/main.nf
@@ -18,11 +18,11 @@
         'biocontainers/glimpse-bio:2.0.0--hf340a29_0' }"
 
     input:
-        tuple val(meta) , path(input), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map)
+        tuple val(meta) , path(input), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map), path(bamlist)
         tuple val(meta2), path(fasta_reference), path(fasta_reference_index)
 
     output:
-        tuple val(meta), path("*.{vcf,bcf,bgen}"), emit: phased_variants
+        tuple val(meta), path("*.{vcf.gz,bcf,bgen}"), emit: phased_variants
         tuple val(meta), path("*.txt.gz")        , emit: stats_coverage, optional: true
         path "versions.yml"                      , emit: versions
 
@@ -42,14 +42,19 @@
     def output_region_cmd     = output_region       ? "--output-region $output_region": ""
 
     def input_bam             = input.any { it.extension in ["cram","bam"]}
+    def input_list            = bamlist ? true : false
 
     """
-    if $input_bam
+    if $input_list
     then
-        ls -1 | grep '\\.cram\$\\|\\.bam\$' > all_bam.txt
-        input_command="--bam-list all_bam.txt"
+        input_command="--bam-list $bamlist"
     else
-        input_command="--input-gl $input"
+        if $input_bam
+        then
+            input_command="--bam-file $input"
+        else
+            input_command="--input-gl $input"
+        fi
     fi
 
     GLIMPSE2_phase \\

************************************************************
