Changes in module 'nf-core/glimpse2/phase'
--- modules/nf-core/glimpse2/phase/main.nf
+++ modules/nf-core/glimpse2/phase/main.nf
@@ -22,7 +22,7 @@
         tuple val(meta2), path(fasta_reference), path(fasta_reference_index)
 
     output:
-        tuple val(meta), path("*.{vcf,bcf,bgen}"), emit: phased_variants
+        tuple val(meta), path("*.{vcf.gz,bcf,bgen}"), emit: phased_variants
         tuple val(meta), path("*.txt.gz")        , emit: stats_coverage, optional: true
         path "versions.yml"                      , emit: versions
 

************************************************************
