Changes in module 'nf-core/bcftools/annotate'
--- modules/nf-core/bcftools/annotate/main.nf
+++ modules/nf-core/bcftools/annotate/main.nf
@@ -8,7 +8,7 @@
         'biocontainers/bcftools:1.18--h8b25389_0' }"
 
     input:
-    tuple val(meta), path(input), path(index), path(annotations), path(annotations_index), path(header_lines)
+    tuple val(meta), path(input), path(index), path(annotations), path(annotations_index), path(header_lines), path(rename_chr)
 
     output:
     tuple val(meta), path("*.{vcf,vcf.gz,bcf,bcf.gz}"), emit: vcf
@@ -18,10 +18,11 @@
     task.ext.when == null || task.ext.when
 
     script:
-    def args    = task.ext.args ?: ''
+    def args    = task.ext.args   ?: ''
     def prefix  = task.ext.prefix ?: "${meta.id}"
-    def header_file = header_lines ? "--header-lines ${header_lines}" : ''
-    def annotations_file = annotations ? "--annotations ${annotations}" : ''
+    def header_file      = header_lines ? "--header-lines ${header_lines}" : ''
+    def annotations_file = annotations  ? "--annotations ${annotations}"   : ''
+    def rename_chr_cmd   = rename_chr   ? "--rename-chrs ${rename_chr}"    : ''
     def extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :
                     args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
                     args.contains("--output-type z") || args.contains("-Oz") ? "vcf.gz" :
@@ -34,6 +35,7 @@
         $args \\
         $annotations_file \\
         $header_file \\
+        $rename_chr_cmd \\
         --output ${prefix}.${extension} \\
         --threads $task.cpus \\
         $input

************************************************************
