Changes in module 'nf-core/bcftools/concat'
--- modules/nf-core/bcftools/concat/main.nf
+++ modules/nf-core/bcftools/concat/main.nf
@@ -21,11 +21,14 @@
     def args = task.ext.args   ?: ''
     prefix   = task.ext.prefix ?: "${meta.id}"
     """
+
+    ls -1v ${vcfs} > order_files.txt
+
     bcftools concat \\
         --output ${prefix}.vcf.gz \\
         $args \\
         --threads $task.cpus \\
-        ${vcfs}
+        -f order_files.txt
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
